src/sqlfluff/dialects/dialect_ansi.py

```diff
@@ -17,11 +17,20 @@
                       OneOf, Delimited, Bracketed, AnyNumberOf, Ref,
                       Anything, LambdaSegment, Indent, Dedent)
 from .base import Dialect
+from .ansi_keywords import ansi_keywords
 
 
 ansi_dialect = Dialect('ansi')
 
 
+anti_template = []
+for n in ansi_keywords.split('\n'):
+    n = n.strip()
+    name = n[0].upper() + n[1:].lower() + 'KeywordSegment'
+    ansi_dialect.add(**{name: KeywordSegment.make(n.lower())})
+    anti_template += [n.upper()]
```

Comment:
I think this is a good approach, and makes the code more readable.

*however*, three issues I think we need to deal with.

## Reserved Keywords

I think what we're doing here is effectively reserving all keywords, and in many cases not all are reserved, and can be used quite productively for variables. Many datastores may use `INDEX` as a column name and I don't think it should raise a `ParseError` if that is the case. I think we should definitely distinguish between *reserved* and *non-reserved* keywords here.

The `anti_template` list, should only contain *reserved* keywords, as these are the only ones which cannot be used under any circumstances. For *non-reserved* keywords, I'm fine auto-generating them, but I don't think they should be in the anti-template.

## Dialects

The current structure of this doesn't allow configuration for different dialects very easily, and in particular when it comes to which keywords are reserved and which aren't, there are quite big differences between dialects [see here for an example](https://www.postgresql.org/docs/8.1/sql-keywords-appendix.html).

I'm open to different ways of doing this if you've got a better idea, but my initial suggestion would actually be to have a few methods on the base `Dialect` class:
- `add_keywords`, which takes an iterable of keywords, and boolean `reserved`. For the ansi dialect, this would be called twice, once with an iterable of reserved words and once with an iterable of the non-reserved words. For other dialects this can be used to add *additional* keywords if required.
- `remove_keywords`, similar to above, but removes keywords altogether. Would be used on dialects which used fewer keywords. Arguably this won't be used much initially, but will be required as the tool matures.
- `alter_keywords`, similar to above, but uses the `reserved` keyword to reassign keywords as reserved or not. I expect that this would be used extensively in defining new dialects where the definitions of which are reserved and which not is different.
- `get_keywords`, this just takes a keyword argument of `reserved`. For generating the anti-pattern we would then use this method, so that for sub dialects, they automatically get the right exclusions without having to redefine `NakedIdentifierSegment` in every sub-dialect.

## Defining the `NakedIdentifierSegment`

Given the suggestion above, the `NakedIdentifierSegment` would be dynamic, and for each dialect it would use whichever keywords were reserved in that dialect.

As above, I'm open to other suggestions, but my first idea would be to allow functions to also be passed to the dialect when adding grammar elements. If a `Segment` or `Grammar` is passed then it behaves as currently, but if something `callable` is passed, then the `callable` is called *with the dialiect itself as an argument*. The `callable` would then have access to the `get_keywords` method above, and so could use it to generate the `anti_template`, in a way which is resilient to redefining available keywords without having to redefine `NakedIdentifierSegment` in every new dialect.

Reply:
The idea of additional functions in the dialect to be able to add, remove and alter keywords is very good. It might be cleaner (and especially allows awesome Linter rules) to add a NonReservedKeyword Segment to the SingleIdentifierGrammar to achieve that non-reserved keywords can be used as identifiers. The rule in mind is basically a warning whenever a non-reserved key word is used a column name. 

I agree that the current action to raise parser errors is very restrictive. I had to change all test cases using column names like rownum, id, name, text, date, day and table names like table_name, result, dummy. 

„If a Segment or Grammar is passed then it behaves as currently, but if something callable is passed, then the callable is called with the dialiect itself as an argument. The callable would then have access to the get_keywords method above, and so could use it to generate the anti_template, in a way which is resilient to redefining available keywords without having to redefine NakedIdentifierSegment in every new dialect.“ This appears really complicated to me and non-intuitive. Maybe because I did not get it. 

My suggestion is rather to update the SingleIdentifierGrammar and NonReservedKeyword whenever a keyword is added, removed, altered using the new functions.

Reply:
I don't think I'm explaining myself very well, I'll make a branch with the functionality that I think could work well, and then you're welcome to expand from there. 😃